File: apps/android/VLinkerOBD/app/src/main/java/com/jslee1972/vlinkerobd/obd/BitFieldExtractor.kt
package com.jslee1972.vlinkerobd.obd

/**
 * Specifies a signal packed at the bit level within a (possibly long, multi-byte) UDS payload —
 * the representation manufacturer telemetry like battery cell voltages or tire pressure PIDs
 * needs, versus the plain byte-arithmetic [PidFormula] used by simple standard PIDs.
 *
 * Bit 0 is the most-significant bit of the first payload byte (the byte right after the
 * mode+PID echo), and bit numbering proceeds MSB-first across successive bytes — the
 * "big-endian"/Motorola bit order used throughout the OBDb community signalsets
 * (https://github.com/OBDb) this schema is modeled on. [signed] applies two's-complement
 * sign extension. Final value = raw * [multiplier] / [divisor] + [offset], clamped to
 * [min]/[max] when given.
 */
data class BitFieldSpec(
    val bitIndex: Int,
    val bitLength: Int,
    val multiplier: Double = 1.0,
    val divisor: Double = 1.0,
    val offset: Double = 0.0,
    val signed: Boolean = false,
    val min: Double? = null,
    val max: Double? = null,
)

object BitFieldExtractor {

    fun extractRaw(bytes: List<Int>, bitIndex: Int, bitLength: Int, signed: Boolean): Long? {
        if (bitLength <= 0 || bitLength > 63 || bitIndex < 0) return null
        val totalBits = bytes.size * 8
        if (bitIndex + bitLength > totalBits) return null

        var raw = 0L
        for (offset in 0 until bitLength) {
            val bitPos = bitIndex + offset
            val byteValue = bytes[bitPos / 8]
            val bitInByte = bitPos % 8
            val bit = (byteValue shr (7 - bitInByte)) and 1
            raw = (raw shl 1) or bit.toLong()
        }

        if (signed) {
            val signBit = 1L shl (bitLength - 1)
            if (raw and signBit != 0L) {
                raw -= (1L shl bitLength)
            }
        }
        return raw
    }

    fun evaluate(bytes: List<Int>, spec: BitFieldSpec): Double? {
        val raw = extractRaw(bytes, spec.bitIndex, spec.bitLength, spec.signed) ?: return null
        var value = raw.toDouble() * spec.multiplier / spec.divisor + spec.offset
        spec.min?.let { value = maxOf(value, it) }
        spec.max?.let { value = minOf(value, it) }
        return value
    }
}

File: apps/ios/VLinkerOBD/BitFieldExtractor.swift
import Foundation

/// Decodes a ``BitFieldSpec`` against response payload bytes. Faithful port of Android's
/// `BitFieldExtractor.kt` — bit 0 is the MSB of the first payload byte, MSB-first across
/// successive bytes (the OBDb community "Motorola" bit order), two's-complement for `signed`.
enum BitFieldExtractor {

    static func extractRaw(bytes: [Int], bitIndex: Int, bitLength: Int, signed: Bool) -> Int64? {
        guard bitLength > 0, bitLength <= 63, bitIndex >= 0 else { return nil }
        let totalBits = bytes.count * 8
        guard bitIndex + bitLength <= totalBits else { return nil }

        var raw: Int64 = 0
        for offset in 0..<bitLength {
            let bitPos = bitIndex + offset
            let byteValue = bytes[bitPos / 8]
            let bitInByte = bitPos % 8
            let bit = (byteValue >> (7 - bitInByte)) & 1
            raw = (raw << 1) | Int64(bit)
        }

        if signed {
            let signBit: Int64 = 1 << (bitLength - 1)
            if raw & signBit != 0 {
                raw -= (Int64(1) << bitLength)
            }
        }
        return raw
    }

    static func evaluate(bytes: [Int], spec: BitFieldSpec) -> Double? {
        guard let raw = extractRaw(bytes: bytes, bitIndex: spec.bitIndex, bitLength: spec.bitLength, signed: spec.signed) else {
            return nil
        }
        var value = Double(raw) * spec.multiplier / spec.divisor + spec.offset
        if let minValue = spec.min { value = max(value, minValue) }
        if let maxValue = spec.max { value = min(value, maxValue) }
        return value
    }
}
